### PR TITLE
Implement proposed SA1028: NoTrailingWhitespace

### LINKED ISSUE #729: Rule proposal: NoTrailingWhitespace
As found [here](https://github.com/AArnott/nerdbank.stylecop.rules/blob/master/src/NerdBank.StyleCop.Rules/NerdBankRules.cs) and described [here](http://blog.nerdbank.net/2008/09/notrailingwhitespace-stylecop-rule-and.html), the NoTrailingWhitespace rule helps avoid code where there is invisible trailing whitespace at the end of a line of code. 

Git and some other tools will sometimes flag when there is trailing whitespace, so this helps detect it early.

This issue tracks porting this rule to StyleCop.Analyzers. Feel free to assign to me.

### PR DESCRIPTION
**Pull request by AArnott**
This includes the analyzer, the code fix provider, and the tests.

Of notable mention is that IIRC the original StyleCop did not notice trailing whitespace at the end of a single line or multi-line comment. I took some extra trouble to fill that gap. I believe this will find *all* trailing whitespace. Although we often want to emulate legacy StyleCop behavior, I believe this is simply filling a bug in the original and I expect especially given the auto fix provider that folks can quickly upgrade their code if necessary when applying this rule.

Fix #729

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs [hunk 1]
```diff
@@ -23,10 +23,7 @@ public abstract partial class DiagnosticVerifier
         /// <returns>
         /// A new instance of the C# analyzer being tested.
         /// </returns>
-        protected virtual DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
-        {
-            return null;
-        }
+        protected abstract DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs]: A very good change, but I think it would be better to make a separate pull request for it.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        /// <inheritdoc/>
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: This line should be moved outside of the foreach loop for performance reasons.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        /// <summary>
+        /// Removes trailing whitespace.
+        /// </summary>
+        /// <param name="document">The document to be changed.</param>
+        /// <param name="diagnostic">The diagnostic to fix.</param>
+        /// <param name="cancellationToken">The cancellation token associated with the fix action.</param>
+        /// <returns>The transformed document.</returns>
+        private static async Task<Document> RemoveWhitespaceAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: Passing the root as parameter to this method will be more efficient.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var trivia = root.FindTrivia(diagnosticSpan.Start, findInsideTrivia: true);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: Passing the trivia as parameter to this method will be more efficient.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    string oldTriviaContent = trivia.ToFullString();
+                    TextSpan diagnosticSpanWithinTrivia = TextSpan.FromBounds(diagnosticSpan.Start - triviaLocation.SourceSpan.Start, diagnosticSpan.End - triviaLocation.SourceSpan.Start);
+                    newTriviaContent = string.Concat(
+                        oldTriviaContent.Substring(0, diagnosticSpanWithinTrivia.Start),
+                        oldTriviaContent.Substring(diagnosticSpanWithinTrivia.End));
+                    newTrivia = SyntaxFactory.SyntaxTrivia(trivia.Kind(), newTriviaContent);
+                    newRoot = root.ReplaceTrivia(trivia, newTrivia);
+                    break;
+                default:
+                    throw new InvalidOperationException("Unexpected this is.");
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: How will VS2015 handle this exception? Is there a risk that it will disable the analyzer?

Maybe we should go with something like:
```C#
    Debug.Fail("Unexpected this is.");
    newRoot = root;
    break;
```

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    public class SA1028NoTrailingWhitespace : DiagnosticAnalyzer
+    {
+        /// <summary>
+        /// The ID for diagnostics produced by the <see cref="SA1028NoTrailingWhitespace"/> analyzer.
+        /// </summary>
+        public const string DiagnosticId = "SA1028";
+        internal const string Title = "Code must not contain trailing whitespace";
+        internal const string MessageFormat = "Remove trailing whitespace";
+        internal const string Category = "Style";
+
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: Would be nice if you could provide a help link to the GitHub wiki for the project.
(See #768 for an example)

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                        }
+
+                        if (reportWarning)
+                        {
+                            context.ReportDiagnostic(Diagnostic.Create(Rule, trivia.GetLocation()));
+                        }
+
+                        break;
+                    case SyntaxKind.SingleLineCommentTrivia:
+                        var text = root.GetText();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: This can be done outside of the foreach loop, will improve performance.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                        TextSpan trailingWhitespace = FindTrailingWhitespace(text, trivia.Span);
+                        if (!trailingWhitespace.IsEmpty)
+                        {
+                            context.ReportDiagnostic(Diagnostic.Create(Rule, Location.Create(context.Tree, trailingWhitespace)));
+                        }
+
+                        break;
+                    case SyntaxKind.MultiLineCommentTrivia:
+                    case SyntaxKind.SingleLineDocumentationCommentTrivia:
+                        text = root.GetText();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: This can be done outside of the foreach loop, will improve performance.

**Comment by vweijsters** [comment]
:+1: Looks good. I've made a few remarks (mostly performance related)

**Comment by vweijsters** [comment]
We should definitely profile this one to see if there may be performance issues. Its scanning a lot of items each time the syntax tree is modified.

**Comment by vweijsters** [comment]
@AArnott Should this rule also cover trailing whitespace after conditional directives (like `#if (Debug) `)?
I couldn't find any evidence that it does.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs (hunk 1 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs]: Agreed, and it should be done in the same pull request as addressing `GetCSharpCodeFixProvider`.

Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs]: Reverted. I'll be happy to send another pull request.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs (hunk 2 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: Quite right.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs (hunk 3 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: I considered it, but decided not to for a few reasons:

1. I expect Roslyn caches the result after the first call, making subsequent calls cheap anyway.
1. When that delegate is invoked, we're applying a code fix, which is a relatively expensive operation anyway, so this operation that was cheap enough to run thousands of times for lots of analyzers can certainly afford to run one more time as part of a user initiated action to apply a code fix.
1. Another parameter on a method makes it that much harder to call, and makes the method that much less helpful to its caller, since the caller has to obtain a value that the helper method could obtain itself.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs (hunk 4 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: Same answer as passing in the SyntaxRoot.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs (hunk 5 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: Ya, I wasn't sure what was the best way to go here. So I've made the change you suggest.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs (hunk 1 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/DiagnosticVerifier.cs]: @AArnott The analyzers that don't have a code fix should technically extend `DiagnosticVerifier`, not `CodeFixVerifier`.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs (hunk 6 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: I would prefer to see a **SA1028.md** file in this repository, since the wiki does not allow for pull requests.

Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: Done

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs (hunk 7 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: fixed

**Comment by AArnott** [comment]
Good catch on the condition directives. I'm working on adding that now.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                    TextSpan diagnosticSpanWithinTrivia = TextSpan.FromBounds(diagnosticSpan.Start - triviaLocation.SourceSpan.Start, diagnosticSpan.End - triviaLocation.SourceSpan.Start);
+                    newTriviaContent = string.Concat(
+                        oldTriviaContent.Substring(0, diagnosticSpanWithinTrivia.Start),
+                        oldTriviaContent.Substring(diagnosticSpanWithinTrivia.End));
+                    newTrivia = SyntaxFactory.SyntaxTrivia(trivia.Kind(), newTriviaContent);
+                    newRoot = root.ReplaceTrivia(trivia, newTrivia);
+                    break;
+                default:
+                    Debug.Assert(false, "Unexpected this is.");
+                    newRoot = root;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: :bulb: I've seen every manner of horrible handling on the IDE side for this. It may have been fixed, but the sure way to not have it go wrong is to just `return document;` (the input unaltered).

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        /// <summary>
+        /// The ID for diagnostics produced by the <see cref="SA1028NoTrailingWhitespace"/> analyzer.
+        /// </summary>
+        public const string DiagnosticId = "SA1028";
+        private const string Title = "Code must not contain trailing whitespace";
+        private const string MessageFormat = "Remove trailing whitespace";
+        private const string Category = "Style";
+        private const string Description = "There should not be any whitespace at the end of a line of code.";
+        private const string HelpLink = "https://github.com/DotNetAnalyzers/StyleCopAnalyzers/wiki/SA1028";
+        private const bool IsEnabledByDefault = true;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: :bulb: It would make sense (as a separate pull request) to create a new property `AnalyzerConstants.EnabledByDefault` for this, and use it consistently through the project. I created issue #783 for this.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+        private static DiagnosticDescriptor Rule { get; } =
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028NoTrailingWhitespace.cs]: :bulb: This line doesn't follow the pattern used by all the other diagnostics in the project.

**Comment by sharwell** [comment]
I'm hesitant to enable this by default prior to seeing it in action under the profiler, but I do want to get it included. Can you make sure the following items are clearly addressed in the documentation for the rule? For each case, please indicate whether or not the analyzer checks for trailing whitespace in this location, and also whether or not the code fix removes trailing whitespace in this location.

* :question: What is the behavior of the rule in disabled code (e.g. between `#if false` and `#endif`)?
* :question: What is the behavior of the rule within line comments?
* :question: What is the behavior of the rule within documentation comments?
* :question: What is the behavior of the rule within verbatim string literals?

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs (hunk 9 above)
Comment (AArnott) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs]: I'll fix that.

**Comment by AArnott** [comment]
@sharwell I'm pleased to say that all your questions are already answered by tests I've written, with the exception of the first one (condition directives) which I'm adding now. But the answers so far, and planned for the first one, is that the analyzer is active in all those contexts. And I can certainly run a perf trace on a project with this turned on to see how much time it adds.

**Comment by sharwell** [comment]
I'm really hoping the IDE will delay evaluation of this rule long enough that it's not constantly appearing at the caret location while typing code...

**Comment by AArnott** [comment]
Good point, @sharwell. I wonder if there's a way to detect when there is an incomplete statement and skip it. No point in reporting trailing whitespace when the line itself is incomplete or wrong in some other way.

**Comment by sharwell** [comment]
We could check if the `CompilationUnitSyntax`'s `ContainsDiagnostics` property is true, but that would cause *all* diagnostics in the file to disappear mid-statement while typing. It also doesn't cover typing in comments.

Best thing right now would probably be to experiment with it and see if it is annoying.

**Comment by AArnott** [comment]
Aside: As usual, any time I'm using VS 2015, I find a bug. This time I found a gross memory leak. Just in working with StyleCopAnalyzers, devenv.exe is taking 1.2GB of RAM. Time to file another bug.

**Comment by AArnott** [comment]
I've tested the performance of SA1028 and it's hard to even get it to show up on traces. And when it does, it's very small. So I think we're OK here. This doesn't surprise me because I was careful to avoid allocations or expensive work as much as possible.

I also tried the experience of simply typing code while this rule was turned on. It doesn't look like it pops up while I'm typing much. I can find it if I pause and look through the error list but otherwise it wasn't noticeable. I recommend we merge this.
